crates/common/trie/flattrie.rs

```diff
@@ -0,0 +1,253 @@
+
+/// A trie implementation that is non recursive, POD and avoids copying and encoding
+/// nodes by providing views into a RLP flat buffer
+#[derive(
+    Default,
+    serde::Serialize,
+    serde::Deserialize,
+    rkyv::Serialize,
+    rkyv::Deserialize,
+    rkyv::Archive,
+    Clone,
+)]
+pub struct FlatTrie {
+    /// Stores a contiguous byte buffer with each RLP encoded node
+    pub data: Vec<u8>,
+    /// Contains the structural information of the MPT
+    pub views: Vec<NodeView>,
+    /// The index of the view for the root of this trie
+    pub root_index: usize,
+    /// Root hash that gets initialized when calling `Self::authenticate`
+    #[serde(skip)]
+    #[rkyv(with = Skip)]
+    root_hash: Option<[u8; 32]>,
+}
+
+/// A view into a particular node
+#[derive(
+    Clone,
+    Copy,
+    serde::Serialize,
+    serde::Deserialize,
+    rkyv::Serialize,
+    rkyv::Deserialize,
+    rkyv::Archive,
+)]
+pub struct NodeView {
+    /// Indices to the RLP code of this node over the flat data buffer
+    pub data_range: (usize, usize),
+    /// Handle into this node's childs
+    pub childs: NodeChilds,
+}
+
+/// Contains indices to the `handles` list for each child of a node
+#[derive(
+    Clone,
+    Copy,
+    serde::Serialize,
+    serde::Deserialize,
+    rkyv::Serialize,
+    rkyv::Deserialize,
+    rkyv::Archive,
+)]
+pub enum NodeChilds {
+    /// A leaf node doesn't have any childs
+    Leaf,
+    /// An extension node always has a branch as a child
+    Extension { child: Option<usize> },
+    /// A branch node can have any node type as any of its 16 childs
+    /// TODO: This can be optimized to a bitmap if the data vec is ordered (contiguous childs)
+    Branch { childs: [Option<usize>; 16] },
+}
+
+impl From<&Node> for FlatTrie {
+    fn from(root: &Node) -> Self {
+        let mut trie = FlatTrie::default();
+
+        fn recursive(value: &Node, trie: &mut FlatTrie) {
+            let childs = match value {
+                Node::Branch(node) => {
+                    let mut childs = [None; 16];
+                    for (i, choice) in node.choices.iter().enumerate() {
+                        if let NodeRef::Node(choice, _) = choice {
+                            recursive(&(*choice), trie);
+                            childs[i] = Some(trie.views.len() - 1);
+                        }
+                    }
+                    NodeChilds::Branch { childs }
+                }
+                Node::Extension(node) => {
+                    let mut child = None;
+                    if let NodeRef::Node(child_node, _) = &node.child {
+                        recursive(child_node, trie);
+                        child = Some(trie.views.len() - 1);
+                    }
+                    NodeChilds::Extension { child }
+                }
+                Node::Leaf(_) => NodeChilds::Leaf,
+            };
+
+            let offset = trie.data.len();
+            trie.data.extend(value.encode_to_vec());
```

Comment:
It's probably better to do:
```suggestion
            value.encode(&mut trie.data);
```
That should save a memcpy and a second realloc.